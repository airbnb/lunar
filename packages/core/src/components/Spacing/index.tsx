import React from 'react';
import { StyleBlock } from 'aesthetic';
import withStyles, { WithStylesProps } from '../../composers/withStyles';

export type SpacingRange =
  | 0
  | 0.5
  | 1
  | 1.5
  | 2
  | 2.5
  | 3
  | 3.5
  | 4
  | 4.5
  | 5
  | 5.5
  | 6
  | 6.5
  | 7
  | 7.5
  | 8
  | 8.5
  | 9
  | 9.5
  | 10
  | 10.5
  | 11
  | 11.5
  | 12;

export type Props = {
  /** Apply spacing on all sides. */
  all?: SpacingRange;
  /** Apply spacing on the bottom. */
  bottom?: SpacingRange;
  /** Content to render. */
  children: NonNullable<React.ReactNode>;
  /** Apply spacing on the left and right. */
  horizontal?: SpacingRange;
  /** Render as inline block instead of block. */
  inline?: boolean;
  /** Use padding instead of margin. */
  inner?: boolean;
  /** Apply spacing on the left. */
  left?: SpacingRange;
  /** Apply spacing on the right. */
  right?: SpacingRange;
  /** Base element / component type. Can be a string representing a DOM element. */
  tag?: 'article' | 'div' | 'footer' | 'header' | 'section';
  /** Apply spacing on the top. */
  top?: SpacingRange;
  /** Apply spacing on the top and bottom. */
  vertical?: SpacingRange;
};

function cleanClassName(className: string | number) {
  return String(className).replace('.', 'dot');
}

/** Common component for arbitray layout structure and spacing. */
export class Spacing extends React.Component<Props & WithStylesProps> {
  static defaultProps: Partial<Props> = {
    all: 0,
    bottom: 0,
    horizontal: 0,
    inline: false,
    inner: false,
    left: 0,
    right: 0,
    top: 0,
    vertical: 0,
  };

  render() {
    const {
      cx,
      all,
      bottom,
      children,
      horizontal,
      left,
      tag: TagProp = 'div',
      top,
      inline,
      inner,
      right,
      styles,
      vertical,
    } = this.props;
    const type = inner ? 'inner' : 'outer';
    const classes: any = [];

    if (all) {
      const cleaned = cleanClassName(all);
      classes.push(
        styles[`${type}Top_${cleaned}`],
        styles[`${type}Right_${cleaned}`],
        styles[`${type}Bottom_${cleaned}`],
        styles[`${type}Left_${cleaned}`],
      );
    } else {
      if (vertical) {
        const cleaned = cleanClassName(vertical);
        classes.push(styles[`${type}Top_${cleaned}`], styles[`${type}Bottom_${cleaned}`]);
      } else {
        if (top) {
          const cleaned = cleanClassName(top);
          classes.push(styles[`${type}Top_${cleaned}`]);
        }

        if (bottom) {
          const cleaned = cleanClassName(bottom);
          classes.push(styles[`${type}Bottom_${cleaned}`]);
        }
      }

      if (horizontal) {
        const cleaned = cleanClassName(horizontal);
        classes.push(styles[`${type}Left_${cleaned}`], styles[`${type}Right_${cleaned}`]);
      } else {
        if (left) {
          const cleaned = cleanClassName(left);
          classes.push(styles[`${type}Left_${cleaned}`]);
        }

        if (right) {
          const cleaned = cleanClassName(right);
          classes.push(styles[`${type}Right_${cleaned}`]);
        }
      }
    }

    return (
      <TagProp className={cx(...classes, inline && styles.spacing_inline)}>{children}</TagProp>
    );
  }
}

export default withStyles(({ unit }) => {
  const spacing: { [key: string]: StyleBlock } = {};

  for (let i = 0; i <= 12; i += 0.5) {
    const size = unit * i;

    ['Top', 'Right', 'Bottom', 'Left'].forEach(side => {
      const cleaned = cleanClassName(i);
      spacing[`outer${side}_${cleaned}`] = { [`margin${side}`]: size };
      spacing[`inner${side}_${cleaned}`] = { [`padding${side}`]: size };
    });
  }

  return {
    ...spacing,

    spacing_inline: {
      display: 'inline-block',
    },
  };
})(Spacing);
