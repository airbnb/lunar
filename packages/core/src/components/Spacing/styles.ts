import { StyleBlock } from 'aesthetic';
import { StyleSheet } from '../../hooks/useStyles';

export function cleanClassName(className: string | number) {
  return String(className).replace('.', 'dot');
}

const styleSheet: StyleSheet = ({ unit }) => {
  const spacing: { [key: string]: StyleBlock } = {};

  for (let i = 0; i <= 12; i += 0.5) {
    const size = unit * i;

    ['Top', 'Right', 'Bottom', 'Left'].forEach(side => {
      const cleaned = cleanClassName(i);
      spacing[`outer${side}_${cleaned}`] = { [`margin${side}`]: size };
      spacing[`inner${side}_${cleaned}`] = { [`padding${side}`]: size };
    });
  }

  return {
    ...spacing,

    spacing_inline: {
      display: 'inline-block',
    },
  };
};

export default styleSheet;

export { styleSheet };
