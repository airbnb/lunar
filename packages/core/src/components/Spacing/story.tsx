import React from 'react';
import StatusLabel from '../StatusLabel';
import Spacing from '.';

export default {
  title: 'Core/Spacing',
  parameters: {
    inspectComponents: [Spacing],
  },
};

export function applyTopMargin() {
  return (
    <Spacing top={3}>
      <StatusLabel>Content</StatusLabel>
    </Spacing>
  );
}

applyTopMargin.story = {
  name: 'Apply top margin.',
};

export function applyRightMargin() {
  return (
    <Spacing right={3}>
      <StatusLabel>Content</StatusLabel>
    </Spacing>
  );
}

applyRightMargin.story = {
  name: 'Apply right margin.',
};

export function applyBottomMargin() {
  return (
    <Spacing bottom={3}>
      <StatusLabel>Content</StatusLabel>
    </Spacing>
  );
}

applyBottomMargin.story = {
  name: 'Apply bottom margin.',
};

export function applyLeftMargin() {
  return (
    <Spacing left={3}>
      <StatusLabel>Content</StatusLabel>
    </Spacing>
  );
}

applyLeftMargin.story = {
  name: 'Apply left margin.',
};

export function applyTopAndBottomVerticalMargin() {
  return (
    <Spacing vertical={3}>
      <StatusLabel>Content</StatusLabel>
    </Spacing>
  );
}

applyTopAndBottomVerticalMargin.story = {
  name: 'Apply top and bottom (vertical) margin.',
};

export function applyLeftAndRightHorizontalMargin() {
  return (
    <Spacing horizontal={3}>
      <StatusLabel>Content</StatusLabel>
    </Spacing>
  );
}

applyLeftAndRightHorizontalMargin.story = {
  name: 'Apply left and right (horizontal) margin.',
};

export function applyMarginOnAllSides() {
  return (
    <Spacing all={3}>
      <StatusLabel>Content</StatusLabel>
    </Spacing>
  );
}

applyMarginOnAllSides.story = {
  name: 'Apply margin on all sides.',
};

export function usePaddingInsteadOfMargin() {
  return (
    <Spacing inner all={3}>
      <StatusLabel>Content</StatusLabel>
    </Spacing>
  );
}

usePaddingInsteadOfMargin.story = {
  name: 'Use padding instead of margin.',
};

export function displayInline() {
  return (
    <>
      <Spacing inline all={0.5}>
        <StatusLabel>Content 1</StatusLabel>
      </Spacing>
      <Spacing inline all={0.5}>
        <StatusLabel>Content 2</StatusLabel>
      </Spacing>
    </>
  );
}

displayInline.story = {
  name: 'Display inline.',
};
