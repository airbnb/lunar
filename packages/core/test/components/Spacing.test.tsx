import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import Spacing from '../../src/components/Spacing';

describe('<Spacing />', () => {
  it('renders with a side', () => {
    const wrapper = shallowWithStyles(<Spacing top={1}>Content</Spacing>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders with many sides', () => {
    const wrapper = shallowWithStyles(
      <Spacing top={1} bottom={3} left={2} right={4}>
        Content
      </Spacing>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders as inline', () => {
    const wrapper = shallowWithStyles(
      <Spacing vertical={2} inline>
        Content
      </Spacing>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders with padding', () => {
    const wrapper = shallowWithStyles(
      <Spacing all={2} inner>
        Content
      </Spacing>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders a <div /> by default', () => {
    const wrapper = shallowWithStyles(<Spacing>Content</Spacing>);

    expect(wrapper.type()).toEqual('div');
  });

  it('renders with the passed in tag type', () => {
    const wrapper = shallowWithStyles(<Spacing tag="footer">Footer content</Spacing>);

    expect(wrapper.type()).toEqual('footer');
  });
});
