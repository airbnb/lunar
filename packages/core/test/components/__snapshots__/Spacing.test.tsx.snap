// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Spacing /> renders as inline 1`] = `
<div
  className="outerTop_2_1vk19cb-o_O-outerBottom_2_p03egf-o_O-spacing_inline_36rlri"
>
  Content
</div>
`;

exports[`<Spacing /> renders with a side 1`] = `
<div
  className="outerTop_1_klarpw"
>
  Content
</div>
`;

exports[`<Spacing /> renders with many sides 1`] = `
<div
  className="outerTop_1_klarpw-o_O-outerBottom_3_152qbzi-o_O-outerLeft_2_1mqc21n-o_O-outerRight_4_h8tpom"
>
  Content
</div>
`;

exports[`<Spacing /> renders with padding 1`] = `
<div
  className="innerTop_2_121xeo4-o_O-innerRight_2_c0l5kf-o_O-innerBottom_2_1ouw2w0-o_O-innerLeft_2_18jukes"
>
  Content
</div>
`;
